File: lib/ios/Crypto/BytesUtils_.swift
//
//  BytesUtils.swift
//
//  Created by Yonatan Rimon on 13/05/2020.
//  Copyright © 2020 iGates. All rights reserved.
//

import Foundation

// Decoding: String(data: bytes, encoding: .ascii)
let STRINGS = ["id" : [UInt8]("IdentityKey".data(using: .ascii)!),
               "id_com" : [UInt8]("IdentityCommitment".data(using: .ascii)!),
               "master0" : [UInt8]("DeriveMasterFirstKey".data(using: .ascii)!),
               "master" : [UInt8]("DeriveMasterKey".data(using: .ascii)!),
               "ddaykey" : [UInt8]("DeriveDayKey".data(using: .ascii)!),
               "dverif" : [UInt8]("DeriveVerificationKey".data(using: .ascii)!),
               "depoch" : [UInt8]("DeriveEpoch".data(using: .ascii)!),
               "verifkey" : [UInt8]("VerificationKey".data(using: .ascii)!)]

class BytesUtils {
    
    // Int to bytes array
    class func numToBytes(num: Int, numBytes: Int) -> [UInt8] {
        var res:[UInt8] = Array()
        var _number:Int = num
        for _ in 0 ..< numBytes {
            res.append(UInt8( _number&0xff ))
            _number >>= 8 //shift 8 times from left to right
        }
        // TODO: ASSERT
        // assert num == 0, "Sanity check"
        return res
    }
    
    // hex string to bytes array
    class func hexToBytes(str: String) -> [UInt8] {
        // omit error checking: remove '0x', make sure even, valid chars
        let pairs = BytesUtils.toPairsOfChars(pairs: [], string: str)
        return pairs.map { UInt8($0, radix: 16)! }
    }
    
    
    class func toPairsOfChars(pairs: [String], string: String) -> [String] {
        if string.count == 0 {
            return pairs
        }
        var pairsMod = pairs
        pairsMod.append(String(string.prefix(2)))
        return toPairsOfChars(pairs: pairsMod, string: String(string.dropFirst(2)))
    }

    // pad
    class func pad(array: [UInt8], size: Int) -> [UInt8] {
        // TODO: ASSERT
        //     assert len(array) <= size, "Padded array should not be smaller."
        if size < array.count {
            return [UInt8]()
        }
        let paddingSize = size - array.count
        let paddingArray = [UInt8](repeating: 0, count: paddingSize)

        return array + paddingArray
    }
    
    // xor
    class func xor(_ a: [UInt8],_ b: [UInt8]) -> [UInt8] {
        // TODO: ASSERT
        //     assert len(a) == len(b)      
        return zip(a, b).map{$0^$1}
    }
}

File: lib/android/src/main/java/moh/gov/il/crypto/utilities/BytesUtils.java
package moh.gov.il.crypto.utilities;

/**
 * Created by hagai on 11/05/2020.
 */
public class BytesUtils {

    public static byte[] numToBytes(int num, int numBytes) {

        byte[] result = new byte[numBytes];
        for(int i = 0; i < numBytes; i ++) {
            result[i] = (byte)(num & 0xff);
            num >>= 8;
        }

        assert num == 0 : "Sanity Check";

        return result;
    }

    // TODO: check if this method is correct
    public static byte[] hexToBytes(String s) {

        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

    public static byte[] pad(byte[] array, int size) {

        assert array.length <= size : "Padded array should be smaller";
        byte[] pad = new byte[size - array.length];
        for(int i = 0; i < pad.length; i ++) {
            pad[i] = 0;
        }
        byte[] result = byteConcatenation(array, pad);

        return result;
    }

    // TODO: check if this method is correct
    public static byte[] xor(byte[] first, byte[] second) {

        assert first.length == second.length;

        byte[] result = new byte[first.length];
        int i = 0;
        for(byte b : first) {
            result[i] = (byte) (b ^ second[i++]);
        }
        return result;
    }

    public static final byte[] byteConcatenation(byte[] firstArray, byte[] secondArray) {

        int firstLen = firstArray.length;
        int secondLen = secondArray.length;

        byte[] result = new byte[firstLen + secondLen];
        System.arraycopy(firstArray, 0, result, 0, firstLen);
        System.arraycopy(secondArray, 0, result, firstLen, secondLen);

        return result;
    }
}
